src/osd/osd_types.cc

```diff
@@ -276,32 +276,35 @@ void request_redirect_t::generate_test_instances(list<request_redirect_t*>& o)
 
 void objectstore_perf_stat_t::dump(Formatter *f) const
 {
-  f->dump_unsigned("commit_latency_ms", os_commit_latency);
-  f->dump_unsigned("apply_latency_ms", os_apply_latency);
+  // *_ms values just for compatibility.
+  f->dump_float("commit_latency_ms", os_commit_latency_ns / 1000000.0);
+  f->dump_float("apply_latency_ms", os_apply_latency_ns / 1000000.0);
+  f->dump_unsigned("commit_latency_ns", os_commit_latency_ns);
+  f->dump_unsigned("apply_latency_ns", os_apply_latency_ns);
 }
 
 void objectstore_perf_stat_t::encode(bufferlist &bl) const
 {
   ENCODE_START(1, 1, bl);
-  ::encode(os_commit_latency, bl);
-  ::encode(os_apply_latency, bl);
+  ::encode(os_commit_latency_ns, bl);
+  ::encode(os_apply_latency_ns, bl);
   ENCODE_FINISH(bl);
 }
 
 void objectstore_perf_stat_t::decode(bufferlist::iterator &bl)
 {
   DECODE_START(1, bl);
-  ::decode(os_commit_latency, bl);
-  ::decode(os_apply_latency, bl);
+  ::decode(os_commit_latency_ns, bl);
```

Comment:
this is not backward compatible: `os_commit_latency_ns` and `os_apply_latency_ns` are 64bits integers, while they were 32 bits before this change. that's why `buffer::end_of_buffer` exception is thrown when the test is trying to decode the encoded blob before this change using the new decoder.

Reply:
What is this blob intended for ? Should I fix tests or code ? Where does this blob used ?

Reply:
The objectstore stats get encoded and transmitted as part of MPGStats in OSD::send_pg_stats (via store->get_cur_stats).

That means that if you make this change in an OSD and someone has an older monitor, the monitor will crash when it tries to decode the message.

Increase the version on encoded structures (increment the first `1` in ENCODE_START in ::encode) and they'll still be understood by older daemons as long as fields are only added (not changed in place).  Annoyingly, that means in this instance that the ::encode function needs to output two u32 fields (truncated from os_commit_latency_ns) and then follow those with the two u64 values.  Then in ::decode you would decode two u32 values, then if the message version was >=2 you'd throw them away and decode two u64 values, else you'd stop.

Check out the (many) places in the ceph tree where " if (struct_v " fragments appear to see how that kind of conditional decoding is done.

Reply:
@jcsp Huge thanks. I have changed code by your advice.